group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -368,9 +369,29 @@ public CompletableFuture<HeartbeatResponseData> heartbeat(
             return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
         }
 
-        return FutureUtils.failedFuture(Errors.UNSUPPORTED_VERSION.exception(
-            "This API is not implemented yet."
-        ));
+        if (!isGroupIdNotEmpty(request.groupId())) {
+            return CompletableFuture.completedFuture(new HeartbeatResponseData()
+                .setErrorCode(Errors.INVALID_GROUP_ID.code()));
+        }
+
+        return runtime.scheduleReadOperation("generic-group-heartbeat",
+            topicPartitionFor(request.groupId()),
+            (coordinator, __) -> coordinator.genericGroupHeartbeat(context, request)
```

Comment:
I am wondering whether using a read operation and ignoring the `offset` parameter is safe here. What's your reasoning?

Reply:
i wasn't too sure on this actually. how should we use the offset parameter, should we pass it down and use it to read the `groups` timeline map?

Reply:
as discussed offline, we will keep it as is since it is how the current coordinator behaves.